### ISSUE TITLE
Support for Persian (fa) language

### ISSUE THREAD

**Issue by CodeWithEmad** [apache/superset#29509]
### Bug description

Hi. I wanted to start translating for the Persian language on the project. As mentioned [here](https://superset.apache.org/docs/contributing/howtos#contributing-translations), I created the language directory, added translations, built it for the frontend, and added it to `superset_configs.py` but frontend crashes when changing the language to Persian.

### How to reproduce the bug

I've created fa directory using:
`pybabel init -i superset/translations/messages.pot -d superset/translations -l fa`
and started translating some strings. then I `build-translations` for the forntend and then, added 

```python
LANGUAGES = {
    "en": {"flag": "us", "name": "English"},
    "fr": {"flag": "fr", "name": "French"},
    "fa": {"flag": "ir", "name": "Persian"},
}
```
inside `superset_config.py` and as I expected, I saw the English, French, and Persian languages, and English and French work fine, but after clicking on the Persian language, the frontend raises an error and the top bar doesn't render.
This the error in in console
```
Uncaught                                                                         jed.js:107
TypeError: Cannot read properties of null (reading 'domain')
```

### Screenshots/recordings

https://github.com/apache/superset/assets/49523567/f2a0395e-4d69-4e03-a79f-da55884a5d8b

### Superset version

master / latest-dev

### Python version

3.10

### Node version

18 or greater

### Browser

Chrome

### Additional context

I've checked the superset logs. nothing weird is happening in the logs.

```
superset_worker_beat  | [2024-07-07 16:55:00,038: INFO/MainProcess] Scheduler: Sending due task reports.scheduler (reports.scheduler)
superset_worker       | [2024-07-07 16:55:00,041: INFO/MainProcess] Task reports.scheduler[3b8cfbed-5d9f-4b01-997a-84e5cb1ae2b3] received
superset_worker       | [2024-07-07 16:55:00,042: DEBUG/ForkPoolWorker-1] [stats_logger] (incr) reports.scheduler
superset_worker       | [2024-07-07 16:55:00,068: INFO/ForkPoolWorker-1] Task reports.scheduler[3b8cfbed-5d9f-4b01-997a-84e5cb1ae2b3] succeeded in 0.026297561002138536s: None
superset_app          | 172.23.0.1 - - [07/Jul/2024:16:55:03 +0000] "GET /lang/fa HTTP/1.1" 302 189 "http://localhost:8088/login/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36"
superset_app          | 172.23.0.1 - - [07/Jul/2024:16:55:03 +0000] "GET / HTTP/1.1" 302 223 "http://localhost:8088/login/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36"
superset_app          | 2024-07-07 16:55:03,275:WARNING:root:Class 'werkzeug.local.LocalProxy' is not mapped
superset_app          | 2024-07-07 16:55:03,275:DEBUG:superset.stats_logger:[stats_logger] (incr) welcome
superset_app          | 172.23.0.1 - - [07/Jul/2024:16:55:03 +0000] "GET /superset/welcome/ HTTP/1.1" 302 201 "http://localhost:8088/login/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36"
superset_app          | 2024-07-07 16:55:03,322:WARNING:superset.db_engine_specs:Unable to load SQLAlchemy dialect gsheets: No module named 'google'
superset_app          | 172.23.0.1 - - [07/Jul/2024:16:55:03 +0000] "GET /login/ HTTP/1.1" 200 4002 "http://localhost:8088/login/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36"
```

### Checklist

- [X] I have searched Superset docs and Slack and didn't find a solution to my problem.
- [X] I have searched the GitHub issue tracker and didn't find a similar bug report.
- [X] I have checked Superset's logs for errors and if I found a relevant Python stacktrace, I included it here as text in the "additional context" section.

**Comment by CodeWithEmad** [comment]
This is so cool, @dosu. the `messages.json` has the appropriate structure:

```json
{
  "domain": "superset",
  "locale_data": {
    "superset": {
      "22": [""],
      "": {
        "domain": "superset",
        "plural_forms": "nplurals=2; plural=(n==0 || n==1);",
        "lang": "fa"
      },
```
No, I didn't have `fa` in my `Locale` because nowhere in the docs mentioned this step, so we need to update the docs for translations. Even having 'fa` in my Locale and running npm ci && npm run build didn't do the trick and I still get the same error.

![image](https://github.com/apache/superset/assets/49523567/114004de-891f-43b0-aa41-5d26fb1296a9)

**Comment by CodeWithEmad** [comment]
didn't solve my issue.

**Comment by CodeWithEmad** [comment]
Sorry to bother you here @rusackas. I followed the documentation but encountered this error. Did I miss a step? Maybe there's a limitation for some languages? I should mention that Persian is a right-to-left (RTL) language, but I just wanted to see the changes in the UI.

**Comment by rusackas** [comment]
All good. As you've surmised, [these](https://superset.apache.org/docs/contributing/howtos/#creating-a-new-language-dictionary) are indeed the best/current instructions. As for possibly missing a step here, did you do these two as well?

<img width="974" alt="image" src="https://github.com/apache/superset/assets/812905/eef5b013-3915-4878-90b2-24c3ace57b33">

I wonder if you can compare the resulting .json files with other languages, and see if there are any differences. Or even run it through GPT/Claude/AI and have it look for anything wrong. You might also open the Persian .po file using [Poedit](https://poedit.net/) and see if there are any warnings/errors visible there. 

If you're getting any visible errors, those might be worth sharing here as well.

**Comment by CodeWithEmad** [comment]
Thanks for the response, @rusackas .

> did you do these two as well?

Yes, both `.json` and `.mo` files are created. 

![image](https://github.com/apache/superset/assets/49523567/5ebb933f-3be4-4cfa-842c-0619a0e9a968)


> You might also open the Persian .po file using Poedit and see if there are any warnings/errors visible there.

I'm using PoEdit. No warnings/errors on save. (I haven't translated everything yet. can it be that?)
 
![image](https://github.com/apache/superset/assets/49523567/4442c716-a3e9-477f-a58a-e3b48fe4befd)

I compared jsons for `fa` and `fr`. The only difference is in the translated strings.

![image](https://github.com/apache/superset/assets/49523567/6a65ad20-66e9-4834-b664-a7f2a8f72d6e)

Something is not right here. in the login screen, there's this message: "Enter your login and password below:", but I can't find it anywhere in the codebase. neither it's French translations: "Entrez votre login et mot de passe:"

**Comment by rusackas** [comment]
Thanks for the dependency PR!

I'm not sure about the strings for the login scree, but I would _guess_ those are implemented in Flask-AppBuilder (FAB) and just don't have the `_()` translation wrapper around them. 

Otherwise, it's not yet clear to me what the PR is. If you want to open a draft PR, maybe we can see what (if anything) fails when I pull/run it.

**Comment by CodeWithEmad** [comment]
Sorry, I shared a lot of details in one comment! 

About the Persian translations and Jed issue, I found no workarounds yet. I did everything step by step, but no luck.

Following the "[Creating new language directory](https://superset.apache.org/docs/contributing/howtos/#creating-a-new-language-dictionary)" I found another issue. The script fails in a fresh virtual environment where Jinja2 is not installed. I've added it in #29548

**Comment by rusackas** [comment]
Thanks for the (now merged) PR, that helps a bunch! Indeed, those steps _should_ work for any language. 

As far as the Right to Left (RTL) situation, there was someone who was proposing to do some CSS/template tweaks to support Arabic, but I'm not sure if their effort went anywhere. We should track that down and see if anything took shape :)

> I haven't translated everything yet. can it be that?

That shouldn't be an issue. Most of our translations are partial, so this wouldn't be unusual.

> Something is not right here. in the login screen, there's this message: "Enter your login and password below:", but I can't find it anywhere in the codebase. neither it's French translations: "Entrez votre login et mot de passe:"

I was able to confirm that these strings are part of a view in [FAB](https://github.com/dpgaspar/Flask-AppBuilder). @dpgaspar might have a good means to rectify this, but the solutions that spring to mind are (1) we should make this view a part of our React Single Page App at some point, and (2) in the meantime, we can probably override the FAB jinja template with one in Superset, and then provide our own translations in any/all supported languages.

**Comment by CodeWithEmad** [comment]
Thanks for the extensive comment. I really appreciate it.

> those steps should work for any language.

I'm out of ideas. My last guess is that probably something is wrong with my setup. Although, I did everything from scratch again. I don't mean to keep this open and can close it if you or someone can confirm the steps works for Persian.

> We should track that down and see if anything took shape :)

Awesome. I found out #25258 

> I was able to confirm that these strings are part of a view in FAB.

Awesome man. Thanks. I'll try to find my way in that repo for login page translations but I love the idea of overriding the template.

**Comment by rusackas** [comment]
If you want to open a draft PR with the code you have, maybe I can pull the branch and see if my luck is any different.

Side note, thank you for all the PRs you've opened thus far. Community contributions like this mean everything :D

**Comment by CodeWithEmad** [comment]
That's so nice of you. well, it's not much of a change, just a `.po` file, and `fa` language added to a local in jed package and superset_config which I should remove but I just tried to create what I've got here. https://github.com/apache/superset/pull/29580


 > Community contributions like this mean everything
 
Indeed, I love open source. I'm currently contributing to the openedx project and its related repositories, but I'm starting to like Superset. Maybe one day I can review pull requests. :)

**Comment by tsator** [comment]
Hi!, @CodeWithEmad am facing the same problem that you describe at the beginning, I follow all the steps in the documentation but at the end of the day I am facing the following error

![image](https://github.com/user-attachments/assets/1f9fe583-e43e-4c76-97b4-cf243fca8c94)

Even if I delete the entire translations folder it still works for some languages ​​so I suspect that it is not taking the translations from the translations folder but they are predefined somewhere else.

I can't figure out if you found a solution, but if you have a solution I would appreciate it if you could help me.

**Comment by CodeWithEmad** [comment]
Hi @tsator. Sorry to hear this. Actually, it was fixed after I rebuilt the project.
Did you follow the steps mentioned in https://superset.apache.org/docs/contributing/development/ to setup your dev environment? 
I hope this helps. It looks like [Jed](https://github.com/messageformat/Jed) is trying to find a messages.json file that contains the `domains`. 

To generate the JSON files, navigate to the `superset-frontend` directory. After installing the npm packages using `npm ci`, run the command `npm run build-translation`. This should create the necessary JSON files.
